### PR TITLE
Cuda TensorAccessor

### PR DESCRIPTION
**Pull request by t-vi**
Provide a TensorAccessor-Like interface for CUDA as discussed in #8366.

Compared to TensorAccessor
- the CUDATensorAccessor copies the sizes and strides while on the host (I didn't implement a host indexing function, though) to enable transfer to the device, on the device, `[]` works like for TensorAccessors,
- instantiation is from TensorAccessors in order to allow using `.accessor<..>`. The drawback is that it you cannot use `auto` for the variable declaration, but the alternative would be a cuda-specific `.accessor`-like function,
- there is a PtrTraits argument to enable `__restrict__`,

Example for the intended use:
```
#include <ATen/cuda/CUDATensorAccessor.cuh>
...
template <typename scalar_t>
__global__ void
apply_homography_2d_kernel(cuda::CUDATensorAccessor<scalar_t, 4> dest_a,
			   cuda::CUDATensorAccessor<scalar_t, 4> src_a,
			   cuda::CUDATensorAccessor<float, 2> transform) {
...
}

template <typename scalar_t>
Tensor apply_homography_2d_template(Tensor& res, const Tensor& image, const Tensor& transform) {
  ...
  cuda::CUDATensorAccessor<scalar_t, 4> image_a(image.accessor<scalar_t, 4>());
  cuda::CUDATensorAccessor<scalar_t, 4> res_a(res.accessor<scalar_t, 4>());
  cuda::CUDATensorAccessor<float, 2> transform_a(transform.accessor<float, 2>());
  auto stream = at::cuda::getCurrentCUDAStream();

  apply_homography_2d_kernel<scalar_t>
    <<<grid, block, 0, stream>>>(res_a, image_a, transform_a);
  return res;
}

...
```

I could use a hint where to put a test for this (e.g. doing a plain vanilla matrix multiplication with a custom kernel) and comparing with the aten mm.

### CODE REVIEW

**Comment by ClementPinard** [comment]
Thanks for the PR !

 * The `cuda::CUDATensorAccessor<scalar_t, 4> image_a(image.accessor<scalar_t, 4>())`
seems very tedious and error prone due to having to specify twice the `scalar_t` and `N` parameters. Couldn't we make a constructor directly from a ATen tensor ?

 * Ideally I guess that, as you suggested, a `.cuda_accessor<>()` would be very nice, but maybe it would mean messing up the tensor class with cuda code, which is not necessarily what we want.

 * And last, I'd say that knowing how to use the `RestrictPtrTraits` is not very easy. Maybe have two different accessors ? something like `cuda::CUDATensorRestrictedAccessor<scalar_t, 4>` along with the default one.

Note : these are comment from an end-user point of view (That writes ATen torch Extensions), so I focus on the interface more than the code cleanliness of ease of implementation, so take it for what it's worth, I might be actually clueless about what CUDA practice should be enforced or not :monkey:

**Comment by t-vi** [comment]
Thank you for the feedback.
I completely agree with having to specify type + dimension twice is terrible. The only good part is that mismatches will result in a compiler error.

I'm not sure about adding another type for restricted. If we had a cuda_accessor supporting PointerTraits, I think it would be OK - especially given that it is a bit of an "expert subject" anyway.

@goldsborough (or anyone else) Would you have advice how to cleanly put in a .cuda_accessor method?

I think we could also add ifdefs to make CUDATensorAccessor available on non-cuda tensors (something that is completely lost at the moment - using `.accessor` for a cuda tensor probably is fishy, too -  and and error out at runtime when is_cuda is false.
If that is a solution: Where would be a good file to define CUDATensorAccessor in?

**Comment by ezyang** [comment]
@t-vi Help me understand the delta between this and TensorAccessor. The differences seem to be:

1. For the non-SubAccessor sizes/strides directly inline the struct, so that you can transfer it easily in CUDA
2. It has appropriate `__host__` and `__device__` annotations
3. Uses `PtrTraits` to support `__restrict__`

So, it seems to me that we should do two things:

1. Make it non-CUDA specific. CUDATensorAccessor is just a version of TensorAccessor that stores its fields inline. So we can call it something like PackedTensorAccessor.
2. Then, we just directly support this in the Tensor API, as `packed_accessor`. So now you can avoid going through TensorAccessor
3. You should generalize TensorAccessor to support PtrTraits, and have `__host__` and `__device__` annotations when CUDA is enabled. Then you can use it uniformly for CPU, and for sub-accessors in CUDA.

Does this make sense?

**Comment by ClementPinard** [comment]
To help understand this better, here is what Cuda accessor is derived from, you can see a lot of similarities :
https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCDeviceTensor.cuh

So in a sense, THCDeviceTensor is to THCTensor what CudaAccessor is to Accessor.

I am no expert in CUDA/CPU c++ templating, but I guess the reason why we need a cuda accessor in Aten in addition to regular accessor are the same as why we need a THCDeviceTensor.

**Comment by ClementPinard** [comment]
I think @gchanan would be a great help to understand the subtlety of this problem, as he helped me use the THCDeviceTensor utilities for my ATen custom module.

**Comment by t-vi** [comment]
@ezyang Makes sense. So there is one thing that doesn't work: device code doesn't have IntList functions/operators, so I would need to make size(i) and stride(i) in TensorAccessor bypass IntList's indexing operator. I'm not sure what the reason is to wrap those into IntList rather than accessing `strides_[i]` directly
https://github.com/pytorch/pytorch/blob/576807ce1a2291895cd1301ec6a982df21beb77f/aten/src/ATen/core/TensorAccessor.h#L20-L21
After making that change, it superficially (i.e. for my immediate use case) seems to work.

**Comment by ezyang** [comment]
> So there is one thing that doesn't work: device code doesn't have IntList functions/operators, so I would need to make size(i) and stride(i) in TensorAccessor bypass IntList's indexing operator. I'm not sure what the reason is to wrap those into IntList rather than accessing strides_[i] directly

That seems like a great reason to just not store IntList. It isn't a big deal to inline it, if it makes it more portable.

**Comment by t-vi** [comment]
Interestingly, the members are plain pointers already, it's just the sizes/strides functions that return IntLists (and so probably won't work on the device, I'd not be too concerned) and the size(i)/stride(i) that is changed now changed not to construct an IntList just to access a single item...

File: aten/src/ATen/templates/Tensor.h [hunk 1]
```diff
@@ -211,6 +211,15 @@ struct AT_API Tensor {
...
   TensorAccessor<T,N> accessor() && = delete;
 
+  template<typename T, size_t N, template <typename U> class PtrTraits = DefaultPtrTraits>
+    PackedTensorAccessor<T,N,PtrTraits> packed_accessor() const& {
+    static_assert(N > 0, "accessor is used for indexing tensor, for scalars use *data<T>()");
+    AT_CHECK(dim() == N, "expected ", N, " dims but tensor has ", dim());
+    return PackedTensorAccessor<T,N,PtrTraits>(static_cast<typename PtrTraits<T>::PtrType>(data<T>()),sizes().data(),strides().data());
+  }
+  template<typename T, size_t N,  template <typename U> class PtrTraits = DefaultPtrTraits>
+  PackedTensorAccessor<T,N> packed_accessor() && = delete;
```
Comment (ezyang) [inline aten/src/ATen/templates/Tensor.h]: A little doc saying what this is good for, and how to use PtrTraits, would be nice :)

File: aten/src/ATen/core/TensorAccessor.h [hunk 2]
```diff
@@ -5,45 +5,117 @@
...
+#ifndef AT_HOSTDEVICE
+#ifdef __CUDACC__
+#define AT_HOSTDEVICE __host__ __device__
+#else
+#define AT_HOSTDEVICE
+#endif
+#endif
 
-template<typename T, size_t N>
+template<typename T, size_t N, template <typename U> class PtrTraits = DefaultPtrTraits>
```
Comment (ezyang) [inline aten/src/ATen/core/TensorAccessor.h]: It's a good idea to state what the new design constraints of TensorAccessorBase are (it works transparently on CUDA) to help future refactorers. Basically the stuff we discussed in the PR.

**Comment by ezyang** [comment]
I think the main thing this is testing is a smoketest in the C++ tests, to make sure it actually does work in CUDA and gives reasonable results. Otherwise, looks good.

File: aten/src/ATen/templates/Tensor.h (hunk 1 above)
Comment (t-vi) [reply aten/src/ATen/templates/Tensor.h]: I added a note further up and also in before Tensor.packed_accessor.

**Comment by t-vi** [comment]
The first three CI failures seem to be unrelated (two timeout, one caffe2 that I cannot really connect to the patch). If nothing turns up, I think I incorporated your feedback. Thank you! @ezyang

**Comment by yf225** [comment]
@ezyang Suggestions for this PR?

**Comment by ezyang** [comment]
Was just waiting for CI to pass

**Pull request closed without merging**